### ISSUE TITLE
Remove the IControllerFactory.ReleaseController

### ISSUE THREAD

**Issue by dotnetjunkie** [aspnet/Mvc#3727]
In ASP.NET 5, the `IControllerFactory` isn't responsible anymore for creation controllers. This responsibility is moved on to the `IControllerActivator` abstraction (which is passed in the controller `Type`). Since it now is the `IControllerActivator` that is responsible of creating types, it should also be responsible of making sure it is 'released' or disposed properly. That basically means that the `IControllerActivator` should have a `ReleaseController` method as well and the `IControllerFactory` implementation should delegate calls to `IControllerFactory.ReleaseController` to `IControllerActivator.ReleaseController`.

This however would be pretty awkward, and now we have to equal points to override `ReleaseController`. So a much better design would be to remove `ReleaseController` altogether. A `ReleaseController` method is redundant, even with DI libraries (i.e. Castle Windsor) where object must be released explicitly.

Since `IControllerActivator` is responsible for creating controllers, it should as well be made responsible of disposing/releasing controllers. Most DI libraries don't need this, but the Castle implementation can ensure deterministic release by wrapping the controller into special 'release wrapper' object that itself is registered for disposal at the end of the request. This object can call the container's `Release` method while passing on the controller. This allows the container to decide what has to be done with the controller. In case the controller is scoped and implements `IDisposable`, it can be disposed.

Another point is that with ASP.NET 5, controllers don't have to inherit from `Controller` of `IController` anymore, which is awesome btw. This allows keeping controllers thin and even stateless, which can improve performance. However, in case a user decides to register (some of) his controllers as singleton, the `DefaultControllerFactory` will currently dispose his controllers at the end of the request, while that controller is intended to be reused. In other words, the underlying DI container isn't informed about this and is completely left out of the equation, which is bad, since it is the DI container that should decide whether or not an object is reused or disposed.

So long story short: The `IControllerFactory.ReleaseController` method should be removed.

**Comment by ploeh** [comment]
This is covered in [my article on DI-friendly frameworks](http://blog.ploeh.dk/2014/05/19/di-friendly-framework). In short, **lifetime management is a single responsibility**; this includes not only creation of objects, but also decommissioning.

A framework API is easiest to understand if the object that creates other objects is also responsible for decommissioning them. So, if `IControllerActivator` creates Controllers, it should also release them when it no longer has need for them.

The ASP.NET MVC 1 & 2 `IControllerFactory` interface was *perfect:*

```C#
public interface IControllerFactory
{
    IController CreateController(
        RequestContext requestContext,
        string controllerName);
 
    void ReleaseController(IController controller);
}
```

It doesn't matter what you *call* such an interface, or exactly which parameters are passed; what matters is that this interface clearly describes the lifetime model used by the framework.

As a client developer, when I'm given such an interface, I don't even have to read the documentation; it's *immediately* clear to me how object lifetime is being managed.

> Most frameworks don't need this

I'm not sure that I agree with this. If you have a nested object graph where one of the leaf nodes are disposable, there are only two ways it can go:

- either you dispose of that object when its lifetime is up
- or you let it stick around until its finalizer is run by the garbage collector

The second option is a poor use of resources, and will impact the capacity of the system employing it. Thus, the only safe, efficient choice is to dispose of objects when their lifetimes are over.

For any object with either Transient or Request-bound lifetime, the framework has to provide a deterministic callback where objects can be disposed of.

It's not that Castle Windsor is being difficult in this regard. On the contrary, it's one of only a few DI Containers that actually take this problem seriously. As a contrast, StructureMap has no decommissioning support, so if object graphs contain disposable nodes, the application *will* have resource leaks.

To be clear: making Controllers implement `IDisposable` doesn't address the decommissioning problem, because you may have dependencies of dependencies, and a Controller will not be able to tell if some concrete implementation in the dependency graph needs to be disposed of.

Notice that I consciously use the term *decommissioning* where appropriate. The reason is that even if an object implements `IDisposable`, it may not mean that is should be disposed of. If, for example, that object has Singleton lifetime, it should only be disposed of when the entire application shuts down.

This is a more complex topic than most people seem to realise, but it doesn't have to be that difficult. I should hope that chapter 8 of [my book](http://amzn.to/12p90MG) does a good job of explaining the issues involved.

**Comment by dotnetjunkie** [comment]
>> Most frameworks don't need this

> I'm not sure that I agree with this.

I'm sorry, I didn't make myself very clear. The reason most DI frameworks don't require explicit release, is because they will release objects for you when the scope (in ASP.NET, a web request) ends. So either the infrastructure will dispose this 'scope' because an `Request_Ends` event got raised, or the user will call `Dispose` once on the scope (which is the case with Autofac's `container.BeginLifetimeScope()`) (and this would hopefully be done by the user inside the composition root). 

So releasing objects *IS* required and should not be ignored, but my argument is that a `ReleaseController` method is *never* required, because the DI container's integration package for ASP.NET ensures releasing at the end of the request for us. For the Castle integration package that means a bit more (but not much) work to make sure controllers are released, compared to the other frameworks. But as I explained, implementing this in Castle is actually trivial. So IMO, all containers actually take this problem seriously, but most manage this for you automatically.

**Comment by ploeh** [comment]
Well, I was never a fan of lifetime scopes, but they're functionally equivalent to Resolve/Release. It's the same model, you're only moving the methods around a bit.

"`ReleaseController` is never required" is true only under the assumption that there's a scope that ends when the request ends. The framework doesn't call `ReleaseController`, but it calls `Dispose` on the scope. Isn't the difference purely semantic?

---

That said, I may have been confused by the wording. In order to make this discussion clearer to readers, can we please agree on the following definitions?

- A **Library** is a reusable set of types or functions you can use from a wide variety of applications. The application code initiates communication with the library and invokes it.
- A **Framework** consists of one or more libraries, but the difference is that Inversion of Control applies. The application registers with the framework (often by implementing one or more interfaces), and the framework calls into the application, which may call back into the framework.

In this discussion, ASP.NET is a framework. All DI Containers I'm aware of are libraries.

**Comment by dotnetjunkie** [comment]
> In this discussion, ASP.NET is a framework. All DI Containers I'm aware of are libraries.

I updated my initial question to make this more clear.

> "ReleaseController is never required" is true only under the assumption that there's a scope that ends when the request ends.

In the case of ASP.NET, there is always a scope when the request ends (since the request is the scope, and we can always be notified when this request is about to end). So for ASP.NET the `ReleaseController` method actually *is* redundant. Not only is it redundant, we now have an extra method on an interface that must be implemented when a custom implementation is made, and the current behavior of the ASP.NET framework is plain wrong, which makes it harder to create a custom implementation (it requires overriding both `IControllerFactory` *and* `IControllerActivator`).

**Comment by Eilon** [comment]
This does seem wrong - the contract for creating and destroying controllers should be the same.

**Comment by lodejard** [comment]
I believe there is another issue filed to give both the IControllerFactory and IControllerActivator a create and release method - so I'm going to be closing this one since it's dealing with a recommendation that should continue over there if needed.

To introduce IoC type activation you only need to override the IControllerActivator. The IControllerFactory calls through to it, and also implements the rest of the work to contextualize the instance after it's instantiated. Things like initializing controller and action context properties, etc.

The reason the activator needs release is that *the controllers are not resolved as IoC services by default* but the user may choose to implement IDisposable which must be called after the request is ended. The default IControllerActivator will call dispose explicitly on Release, but an IoC-based activator only needs to resolve the controller in create and does not need to do anything in release. As you've observed - they can be scoped services and the container will dispose them at the appropriate time.

**Comment by dotnetjunkie** [comment]
> I believe there is another issue filed to give both the IControllerFactory and IControllerActivator a create and release method

Link please?

**Comment by davidfowl** [comment]
@lodejard what if we just registered controllers as services as part of the scan? It might makes things a little awkward since it happens during container build up time.

**Comment by ploeh** [comment]
> I believe there is another issue filed

Which one?

> The reason the activator needs release is that the controllers are not resolved as IoC services by default but the user may choose to implement IDisposable

The fundamental reason a decommissioning hook is required isn't particularly to address Controllers that implement `IDisposable`, but to address the common scenario where a Controller has been injected with an object graph, and deep in that object graph exists an object that should be disposed of.

A common architecture is to have a Controller talk to an injected Domain Service, that again talks to an injected Repository:

`FooController` has an injected `IFooService` object. `IFooService` is an interface, but at run-time, the concrete implementation `FooService` is used.

`FooService` depends on `IFooRepository` (another interface), which is injected into it. At run-time, the concrete class `EntityFrameworkFooRepository` is used, and this object in turn contains an injected `DbContext`-derived object.

To summarise, here's the dependency graph:

`FooController -> FooService -> EntityFrameworkFooRepository -> FooContext`

`FooController` only knows about `IFooService`, which is an interface. An interface shouldn't derive from `IDisposable` as [Nicholas Blumhardt already established in 2010](http://cca.codeplex.com/Thread/View.aspx?ThreadId=82987):

> an interface [...] generally shouldn't be disposable. There's no way for the one defining an interface to foresee all possible implementations of it - you can always come up with a disposable implementation of practically any interface.

This means that `FooController` only knows that it has a reference to an interface. The rest of the dependency graph is hidden to it, and it doesn't know if any nodes in this graph are disposable.

The same argument applies to `FooService`. It can't see beyond its `IFooRepository` dependency.

Still, when `FooController` goes out of scope, it's important to dispose of `FooContext`, because otherwise the application will have resource leaks.

The *only* object that knows about the entire dependency graph is the factory that created it. This is the reason that any API responsible for creating an object graph must also have a decommissioning hook (e.g. a `Release` method).

All of this is made even more complicated by the dimension of lifetime management. Some object graphs may contain nodes that are disposable, but which are also thread-safe. When you have such objects, it makes sense to give them Singleton lifetime. The result of that is that even though they implement `IDisposable`, they shouldn't be disposed of after each request.

So you may have dependency graphs where some nodes aren't disposable at all, some nodes are disposable and should be disposed of, and some nodes are disposable, but shouldn't be disposed of yet.

All of this should be of no concern to the rest of the framework that uses Controllers. All it needs to do is to call `Create` when it needs a Controller, and `Release` when it's done with it. The Composer on which the framework calls `Create` and `Release` takes care of the rest.

(For small object graphs it's actually [easy and beneficial to juggle this lifetime management with 'normal' code](http://blog.ploeh.dk/2014/06/03/compile-time-lifetime-matching), but you can always delegate such work to a dedicated DI Container like Simple Injector or Castle Windsor.)

**Comment by dotnetjunkie** [comment]
> the controllers are not resolved as IoC services by default 

This seems to be a design oversight IMO, because controllers are application components and the container (either the built-in one or a custom one) should be in control of the creation of a controller. This has to be, because a user might want to:

 - control the lifetime of a controller (make it singleton for instance)
 - let a controller have multiple constructors (as some DI containers support). Currently even the built-in container supports multiple constructors, but MVC again forces a single constructor; this is inconsistent and confusing.
 - use property or method injection
 - let the controllers be part of the application's object graph so analysis and diagnostics can be performed by the container about the correctness of the configuration.
 - inject a dependency based on the controller it is injected into.

This all isn't possible with the out-of-the-box behavior of MVC, which makes it really confusing for the user. While .NET Core contains its own built-in container, MVC again defines its own 'container' with different behavior.

There is no doubt in my mind that this behavior should be completely removed. MVC should out-of-the-box simply delegate the creation of controllers to the built-in container. I don't see any good argument why MVC should be in control over the creation.

> the user may choose to implement IDisposable which must be called after the request is ended

This is incorrect; Dispose needs to be called when the application is done using the controller, not after each request. If the controller is a singleton, calling Dispose after a request will clearly cause the application to break.
